codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -57,15 +58,30 @@ public DefaultHttp2RemoteFlowController(Http2Connection connection) {
         connection.addListener(new Http2ConnectionAdapter() {
             @Override
             public void onStreamAdded(Http2Stream stream) {
-                // Just add a new flow state to the stream.
-                stream.remoteFlowState(new DefaultFlowState(stream, 0));
+                // If the stream state is not open then the stream is not yet eligible for flow controlled frames and
+                // only requires the ReducedFlowState. Otherwise the full amount of memory is required.
+                stream.remoteFlowState(stream.state() == IDLE ?
```

Comment:
> Do you have a scenario in mind where you might expect to see this?

The use cases discussed in the [http wg archive](https://lists.w3.org/Archives/Public/ietf-http-wg/2014OctDec/0467.html) revolved around flexibility to create priority groups related to how browsers render web pages.  A claim made was the expectation is to be able to work with this using [hundreds or thousands of stream](https://lists.w3.org/Archives/Public/ietf-http-wg/2014OctDec/0475.html) and being able to prioritize them into groups should cut down response times.  The WG also discussed this feature being especially useful for intermediaries and proxy type use cases where utilization may be more important.

> 1) Why is it relevant what memory usage streams in idle state have?

We should strive to keep our required memory down.  The specification allows clients/servers to prioritize streams before flow control state is necessary.  Keeping this state down allows us to potentially keep our resources down for scenarios where clients/servers take advantage of this portion of the specification.

>  2) Once we switch from idle to open we have to allocate a real FlowState and since ReducedFlowState + FlowState > FlowState we will actually use more memory.

This is no longer the case since this PR was opened.  I fully implemented the optimization so the trigger points are correct.  We only allocate what we need when we need it, and make eligible for GC as soon as we know we can.